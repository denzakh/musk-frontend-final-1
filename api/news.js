import axios from 'axios';

const NEWS_API_KEY = process.env.NEWS_API_KEY;
const NEWS_API_URL = 'https://newsapi.org/v2';

export default async function handler(req, res) {
    if (req.method !== 'GET') {
        res.status(405).json({ error: 'Метод не разрешен' });
        return;
    }

    const {
        type = 'articles',
        q,
        category,
        country,
        pageSize,
        page,
        language,
    } = req.query;

    try {
        let url = '';
        let params = {
            apiKey: NEWS_API_KEY,
        };

        if (type === 'sources') {
            // Получение списка источников
            url = `${NEWS_API_URL}/top-headlines/sources`;
            if (category) params.category = category;
            if (language) params.language = language;
            if (country) params.country = country;
        } else {
            // Получение списка статей
            url = `${NEWS_API_URL}/top-headlines`;
            if (q) params.q = q;
            if (category) params.category = category;
            if (country) params.country = country;
            if (pageSize) params.pageSize = pageSize;
            if (page) params.page = page;
        }

        const response = await axios.get(url, { params });

        res.status(200).json(response.data);
    } catch (error) {
        console.error('Ошибка при запросе к NewsAPI:', error.message);
        res.status(500).json({ error: 'Ошибка сервера при получении данных' });
    }
}
