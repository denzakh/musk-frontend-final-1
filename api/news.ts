import axios from 'axios';

const NEWS_API_KEY = process.env.NEWS_API_KEY;
const NEWS_API_URL = 'https://newsapi.org/v2';

interface NewsApiParams {
    apiKey: string;
    q?: string;
    category?: string;
    country?: string;
    pageSize?: number;
    page?: number;
    language?: string;
}

interface NewsApiRequestQuery extends NewsApiParams {
    type?: string;
}

export default async function handler(
    req: { method: string; query: NewsApiRequestQuery },
    res: {
        status: (arg0: number) => {
            (): unknown;
            new (): unknown;
            json: { (arg0: { error: string }): void; new (): unknown };
        };
    }
): Promise<void> {
    if (req.method !== 'GET') {
        res.status(405).json({ error: 'Метод не разрешен' });
        return;
    }

    const {
        type = 'articles',
        q,
        category,
        country,
        pageSize,
        page,
        language,
    } = req.query;

    try {
        let url = '';
        const params: NewsApiParams = {
            apiKey: NEWS_API_KEY || '',
        };

        if (type === 'sources') {
            url = `${NEWS_API_URL}/top-headlines/sources`;
            if (category) params.category = category as string;
            if (language) params.language = language as string;
            if (country) params.country = country as string;
        } else {
            url = `${NEWS_API_URL}/top-headlines`;
            if (q) params.q = q as string;
            if (category) params.category = category as string;
            if (country) params.country = country as string;
            if (pageSize) params.pageSize = Number(pageSize);
            if (page) params.page = Number(page);
        }

        const response = await axios.get(url, { params });

        res.status(200).json(response.data);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (error: any) {
        console.error('Error de NewsAPI:', error.message);
        res.status(500).json({ error: 'Error de servidor' });
    }
}
