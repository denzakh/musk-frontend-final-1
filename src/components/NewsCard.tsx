import { Article } from '../types/news';
import bookmarkImg from '../assets/bookmark.svg';

interface Props {
    article: Article;
    onSave?: (article: Article) => void;
    isFavorite?: boolean;
}

const NewsCard: React.FC<Props> = ({
    article,
    onSave = () => {},
    isFavorite = false,
}) => (
    <div className='relative rounded-xl shadow-md p-4 bg-white dark:bg-gray-800 transition flex flex-col'>
        {article.urlToImage && (
            <img
                src={article.urlToImage}
                alt={article.title}
                className='rounded-lg mb-2'
            />
        )}
        <h2 className='text-xl font-semibold'>{article.title}</h2>
        <p className='text-sm text-gray-600 dark:text-gray-300'>
            {new Date(article.publishedAt).toLocaleDateString()}
        </p>
        <p className='text-sm mt-2'>{article.description}</p>
        <div
            className='pt-4 flex justify-between items-center'
            style={{ marginTop: 'auto' }}
        >
            <a href={article.url} target='_blank' className='text-blue-500'>
                Читать
            </a>
            {onSave && (
                <button
                    onClick={() => onSave(article)}
                    className='cursor-pointer text-yellow-500'
                >
                    {isFavorite ? 'Remove from favorites' : 'Add to Favorites'}
                </button>
            )}
        </div>
        <div className='absolute right-4 -top-1'>
            {isFavorite && (
                <img
                    src={bookmarkImg}
                    alt={''}
                    width={60}
                    height={60}
                    className=''
                />
            )}
        </div>
    </div>
);

export default NewsCard;
