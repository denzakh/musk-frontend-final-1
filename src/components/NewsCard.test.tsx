import { render, screen, fireEvent } from '@testing-library/react';
import { vi } from 'vitest';
import NewsCard from './NewsCard';
import { Article } from '../types/news';

// Моки изображений
vi.mock('../assets/bookmark.svg', () => ({ default: 'mocked-bookmark.svg' }));

const mockOnSave = vi.fn();
const mockArticle: Article = {
    title: 'Test Article',
    description: 'This is a test description.',
    publishedAt: '2025-04-10T00:00:00Z',
    urlToImage: 'https://via.placeholder.com/150',
    url: 'https://example.com',
    source: { id: 'id', name: 'name' },
    author: '',
    content: '',
};

describe('NewsCard', () => {
    it('debe renderizar la tarjeta de noticia correctamente', () => {
        render(<NewsCard article={mockArticle} />);

        expect(screen.getByText(mockArticle.title)).toBeInTheDocument();

        expect(
            screen.getByText(
                new Date(mockArticle.publishedAt).toLocaleDateString()
            )
        ).toBeInTheDocument();

        expect(screen.getByText(mockArticle.description)).toBeInTheDocument();

        expect(screen.getByText('Read the article')).toHaveAttribute(
            'href',
            mockArticle.url
        );

        expect(screen.getByRole('img')).toHaveAttribute(
            'src',
            mockArticle.urlToImage
        );
    });

    it('debe llamar a onSave cuando se haga clic en el botón "Add to Favorites"', () => {
        render(<NewsCard article={mockArticle} onSave={mockOnSave} />);

        const button = screen.getByText('Add to Favorites');
        fireEvent.click(button);

        expect(mockOnSave).toHaveBeenCalledTimes(1);
        expect(mockOnSave).toHaveBeenCalledWith(mockArticle);
    });

    it('debe cambiar el texto del botón a "Remove from favorites" cuando el artículo es favorito', () => {
        render(
            <NewsCard
                article={mockArticle}
                isFavorite={true}
                onSave={mockOnSave}
            />
        );

        const button = screen.getByText('Remove from favorites');
        expect(button).toBeInTheDocument();
    });

    it('debe mostrar el icono de favorito cuando el artículo es favorito', () => {
        render(<NewsCard article={mockArticle} isFavorite={true} />);

        const bookmarkIcon = screen.getByRole('img');
        expect(bookmarkIcon).toHaveAttribute('src', 'mocked-bookmark.svg');
    });

    it('debe mostrar la imagen del artículo si existe', () => {
        render(<NewsCard article={mockArticle} />);

        const image = screen.getByRole('img');
        expect(image).toHaveAttribute('src', mockArticle.urlToImage);
    });

    it('no debe mostrar la imagen si no existe', () => {
        const articleWithoutImage = { ...mockArticle, urlToImage: null };
        render(<NewsCard article={articleWithoutImage} />);

        const image = screen.queryByRole('img');
        expect(image).toBeNull();
    });

    it('snapshot', () => {
        const { asFragment } = render(<NewsCard article={mockArticle} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
