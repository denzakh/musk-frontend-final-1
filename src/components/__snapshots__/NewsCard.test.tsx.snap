// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewsCard > snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative rounded-xl shadow-md p-4 bg-white dark:bg-gray-800 transition flex flex-col"
  >
    <img
      alt="Test Article"
      class="rounded-lg mb-2"
      src="https://via.placeholder.com/150"
    />
    <h2
      class="text-xl font-semibold"
    >
      Test Article
    </h2>
    <p
      class="text-sm text-gray-600 dark:text-gray-300"
    >
      10.04.2025
    </p>
    <p
      class="text-sm mt-2"
    >
      This is a test description.
    </p>
    <div
      class="pt-4 flex justify-between items-center"
      style="margin-top: auto;"
    >
      <a
        class="text-blue-500 dark:text-blue-300"
        href="https://example.com"
        target="_blank"
      >
        Read the article
      </a>
      <button
        class="cursor-pointer text-blue-500 dark:text-blue-300 "
      >
        Add to Favorites
      </button>
    </div>
    <div
      class="absolute right-4 -top-1"
    />
  </div>
</DocumentFragment>
`;
